ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,436 @@
 
 #include "bcmath.h"
 #include "private.h"
+#include "convert.h"
 #include <stdbool.h>
 #include <stddef.h>
 #include <string.h>
 #include "zend_alloc.h"
 
+static const BC_VECTOR POW_10_LUT[9] = {
+	1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000
+};
 
-/* Some utility routines for the divide:  First a one digit multiply.
-   NUM (with SIZE digits) is multiplied by DIGIT and the result is
-   placed into RESULT.  It is written so that NUM and RESULT can be
-   the same pointers.  */
-
-static void _one_mult(unsigned char *num, size_t size, int digit, unsigned char *result)
+/*
+ * Use this function when the number of array elements in divisor is 1, that is, when divisor is not divided into multiple array elements.
+ * The algorithm can be simplified if divisor is not divided, and this function is an optimization of bc_standard_div.
+ */
+static inline void bc_fast_div(
+	BC_VECTOR *numerator_vector, size_t numerator_arr_size, BC_VECTOR divisor_vector, BC_VECTOR *quot_vector, size_t quot_arr_size)
 {
-	size_t carry, value;
-	unsigned char *nptr, *rptr;
+	size_t numerator_top_index = numerator_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+	for (size_t i = 0; i < quot_arr_size - 1; i++) {
+		if (numerator_vector[numerator_top_index - i] < divisor_vector) {
+			quot_vector[quot_top_index - i] = 0;
+			/* numerator_vector[numerator_top_index - i] is always a smaller number than divisor_vector. Therefore, there will be no overflow. */
+			numerator_vector[numerator_top_index - i - 1] += numerator_vector[numerator_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+			continue;
+		}
+		quot_vector[quot_top_index - i] = numerator_vector[numerator_top_index - i] / divisor_vector;
+		numerator_vector[numerator_top_index - i] -= divisor_vector * quot_vector[quot_top_index - i];
+		numerator_vector[numerator_top_index - i - 1] += numerator_vector[numerator_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+		numerator_vector[numerator_top_index - i] = 0;
+	}
+	/* last */
+	quot_vector[0] = numerator_vector[0] / divisor_vector;
+}
 
-	if (digit == 0) {
-		memset(result, 0, size);
-	} else {
-		if (digit == 1) {
-			memcpy(result, num, size);
-		} else {
-			/* Initialize */
-			nptr = (unsigned char *) (num + size - 1);
-			rptr = (unsigned char *) (result + size - 1);
-			carry = 0;
-
-			while (size-- > 0) {
-				value = *nptr-- * digit + carry;
-				*rptr-- = value % BASE;
-				carry = value / BASE;
-			}
+/*
+ * Used when the number of elements in the divisor vector is 2 or more.
+ */
+static inline void bc_standard_div(
+	BC_VECTOR *numerator_vector, size_t numerator_arr_size,
+	BC_VECTOR *divisor_vector, size_t divisor_arr_size, size_t divisor_len,
+	BC_VECTOR *quot_vector, size_t quot_arr_size
```

Comment:
Do these `arr` parameters refer to the length of the vector?
```suggestion
	BC_VECTOR *numerator_vector, size_t numerator_vector_size,
	BC_VECTOR *divisor_vector, size_t divisor_vector_size, size_t divisor_len,
	BC_VECTOR *quot_vector, size_t quot_vector_size
```

Reply:
If I'm not misunderstanding your question, yes.

However, the macro `BC_VECTOR_SIZE` means how many digits can be stored in the vector, so I intentionally avoid using the expression "vector size" as the array size.

Reply:
Maybe use `numerator_vector_len` then?

Reply:
Ah, okay I get my issue, the `numerator_vector` is an array of vector**s**,

Please make the initial argument plural: i.e. `numerator_vectors` which follows a somewhat standard nomenclature.

Reply:
Fixed
https://github.com/php/php-src/pull/14660/commits/e0b83adc223ae609061e031e27d80d554b385466